pkg/restore/log_client.go

```diff
@@ -0,0 +1,929 @@
+			if !ok {
+				log.Error("[restoreFromPuller] can't get table info from dom by tableID",
+					zap.String("item table", item.Table),
+					zap.String("table", table),
+					zap.String("item schema", item.Schema),
+					zap.String("schema", schema),
+					zap.Int64("backup table id", tableID),
+					zap.Int64("restore table id", newTableID),
+				)
+				return errors.Trace(err)
+			}
+			// reload
+			l.tableBuffers[tableID].ReloadMeta(newTableInfo)
+			log.Debug("reload table meta for table",
+				zap.String("item table", item.Table),
+				zap.String("table", table),
+				zap.String("item schema", item.Schema),
+				zap.String("schema", schema),
+				zap.Int64("backup table id", tableID),
+				zap.Int64("restore table id", newTableID),
+			)
+
+		case cdclog.RowChanged:
+			err := l.tableBuffers[tableID].Append(item)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			if l.tableBuffers[tableID].ShouldApply() {
+				err = l.applyKVChanges(ctx, tableID)
+				if err != nil {
+					return errors.Trace(err)
+				}
+			}
+		}
+	}
+}
+
+func (l *LogClient) restoreTables(ctx context.Context, dom *domain.Domain) error {
+	// 1. decode cdclog with in ts range
+	// 2. dispatch cdclog events to table level concurrently
+	// 		a. encode row changed files to kvpairs and ingest into tikv
+	// 		b. exec ddl
+	log.Debug("start restore tables")
+	workerPool := utils.NewWorkerPool(l.concurrencyCfg.Concurrency, "table log restore")
+	eg, ectx := errgroup.WithContext(ctx)
+	for tableID, puller := range l.eventPullers {
+		pullerReplica := puller
+		tableIDReplica := tableID
+		workerPool.ApplyOnErrorGroup(eg, func() error {
+			return l.restoreTableFromPuller(ectx, tableIDReplica, pullerReplica, dom)
+		})
+	}
+	return eg.Wait()
+}
+
+// RestoreLogData restore specify log data from storage.
+func (l *LogClient) RestoreLogData(ctx context.Context, dom *domain.Domain) error {
+	// 1. Retrieve log data from storage
+	// 2. Find proper data by TS range
+	// 3. Encode and ingest data to tikv
+
+	// parse meta file
+	data, err := l.restoreClient.storage.Read(ctx, metaFile)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	err = json.Unmarshal(data, l.meta)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	if l.startTS > l.meta.GlobalResolvedTS {
+		return errors.Errorf("start ts:%d is greater than resolved ts:%d",
+			l.startTS, l.meta.GlobalResolvedTS)
+	}
+	if l.endTs > l.meta.GlobalResolvedTS {
+		log.Info("end ts is greater than resolved ts,"+
+			" to keep consistency we only recover data until resolved ts",
+			zap.Uint64("end ts", l.endTs),
+			zap.Uint64("resolved ts", l.meta.GlobalResolvedTS))
+		l.endTs = l.meta.GlobalResolvedTS
+	}
+
+	// collect ddl files
+	ddlFiles, err := l.collectDDLFiles(ctx)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// execute create ddl first, because we need tableInfo to create kvEncoder
+	// we will filter all create ddl in later eventPuller
+	err = l.doCreateDDLJob(ctx, ddlFiles)
```

Comment:
I think there maybe some problems here. If there are multi create/drop DDLs with the same table name in ddlFiles, execute all the create ddl make fails or cause unexpected errors

Reply:
You are right, I removed these create logic, and change the logic into `ReloadMeta`. 
every time after `ReloadMeta`, it will set kvEncoder to nil. and in `Append` if it found kvEncoder is nil, it will recreate encoder for next row.